File: tespitsozluk-web/lib/reserved-usernames.ts
/**
 * Backend `ReservedUsernames` ile aynı mantık — güncellerken senkron tutun.
 *
 * a) tr-TR ile küçük harf
 * b) Türkçe harfleri Latin karşılıklarına
 * c) Harf/rakam dışı karakterleri kaldır
 */

const FORBIDDEN_SUBSTRINGS: readonly string[] = [
  "tespitsozluk",
  "admin",
  "yonetici",
  "moderator",
  "destek",
  "support",
  "anonim",
  "anonymous",
  "sistem",
  "system",
  "resmi",
  "official",
  "root",
  "staff",
  "webmaster",
]

const MOD_EXACT = "mod"

function mapTurkishLatin(c: string): string {
  if (c.length !== 1) return c
  switch (c) {
    case "ç":
      return "c"
    case "ğ":
      return "g"
    case "ı":
    case "i":
      return "i"
    case "ö":
      return "o"
    case "ş":
      return "s"
    case "ü":
      return "u"
    default:
      return c
  }
}

/** Kayıt / rezerve kontrolü için normalize (backend ile eşleşmeli). */
export function normalizeForReservedCheck(input: string): string {
  const trimmed = input.trim()
  if (!trimmed) return ""

  let out = ""
  for (const ch of trimmed) {
    const lower = ch.toLocaleLowerCase("tr-TR")
    const mapped = mapTurkishLatin(lower)
    const code = mapped.codePointAt(0)!
    const isAsciiLetter = code >= 0x61 && code <= 0x7a
    const isDigit = code >= 0x30 && code <= 0x39
    if (isAsciiLetter || isDigit) out += mapped
  }

  return out
}

export function isReservedNickname(trimmed: string): boolean {
  const n = normalizeForReservedCheck(trimmed)
  if (!n) return false

  if (n === MOD_EXACT) return true

  for (const fragment of FORBIDDEN_SUBSTRINGS) {
    if (n.includes(fragment)) return true
  }

  return false
}

File: TespitSozluk.API/Helpers/ReservedUsernames.cs
using System.Globalization;
using System.Text;

namespace TespitSozluk.API.Helpers;

/// <summary>
/// Yasaklı nickname kontrolü: Türkçe harf ve özel karakter varyasyonlarını normalize edip kök kelimelerle eşleştirir.
/// </summary>
public static class ReservedUsernames
{
    public const string ReservedMessage =
        "Bu kullanıcı adı sistem tarafından rezerve edilmiştir ve alınamaz.";

    /// <summary>
    /// Kısa token yalnızca tam eşleşmede yasak ("modern" gibi yanlış pozitifleri önlemek için alt dize aranmaz).
    /// </summary>
    private const string ModExact = "mod";

    /// <summary>
    /// Normalize edilmiş kullanıcı adında geçmesi yasak kökler (tam kelime veya içerik olarak).
    /// </summary>
    private static readonly string[] ForbiddenSubstrings =
    [
        "tespitsozluk",
        "admin",
        "yonetici",
        "moderator",
        "destek",
        "support",
        "anonim",
        "anonymous",
        "sistem",
        "system",
        "resmi",
        "official",
        "root",
        "staff",
        "webmaster",
    ];

    /// <summary>
    /// a) Küçük harf (TR kültürü: I/İ tutarlılığı), b) Türkçe harfleri Latin karşılıklarına,
    /// c) Harf ve rakam dışındaki tüm karakterleri kaldırır (nokta, tire, alt çizgi vb.).
    /// </summary>
    public static string NormalizeForReservedCheck(string? input)
    {
        if (string.IsNullOrWhiteSpace(input))
            return string.Empty;

        var tr = CultureInfo.GetCultureInfo("tr-TR");
        var sb = new StringBuilder(input.Trim().Length);

        foreach (var ch in input.Trim())
        {
            var lower = char.ToLower(ch, tr);
            var mapped = MapTurkishLatin(lower);
            if (mapped is >= 'a' and <= 'z' || char.IsAsciiDigit(mapped))
                sb.Append(mapped);
        }

        return sb.ToString();
    }

    private static char MapTurkishLatin(char c) =>
        c switch
        {
            'ç' => 'c',
            'ğ' => 'g',
            'ı' => 'i',
            'ö' => 'o',
            'ş' => 's',
            'ü' => 'u',
            'i' => 'i',
            _ => c,
        };

    public static bool IsReserved(string? username)
    {
        var n = NormalizeForReservedCheck(username);
        if (n.Length == 0)
            return false;

        if (string.Equals(n, ModExact, StringComparison.Ordinal))
            return true;

        foreach (var fragment in ForbiddenSubstrings)
        {
            if (n.Contains(fragment, StringComparison.Ordinal))
                return true;
        }

        return false;
    }
}
